### PR TITLE
[native] Add operational endpoints for connector cache cleanup

### PR DESCRIPTION
**Pull request by tanjialiang**
Adding HTTP operational endpoints for native server for debugging/maintenance purposes.
Currently added:
* HiveConnector file handle cache stats get
* HiveConnector file handle cache clearing

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by xiaoxmeng — COMMENTED** [review]
@tanjialiang thanks for the change. Have suggestions. Also can we have a PrestoServer test which starts a server with http service so we can test it? Thanks!

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 1]
```diff
@@ -81,6 +82,28 @@ void sendOkResponse(proxygen::ResponseHandler* downstream, const json& body) {
...
+      .status(http::kHttpOk, "OK")
+      .header(
+          proxygen::HTTP_HEADER_CONTENT_TYPE, http::kMimeTypeApplicationJson)
+      .body(body)
+      .sendWithEOM();
+}
+void sendErrorResponse(
+    proxygen::ResponseHandler* downstream,
+    const std::string& error = "",
+    int status = http::kHttpInternalServerError) {
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:98]: good catch. I'll refactor it in this PR.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 2]
```diff
@@ -81,6 +82,28 @@ void sendOkResponse(proxygen::ResponseHandler* downstream, const json& body) {
...
+  static const int kMaxStatusSize = 1024;
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:99]: It's from the original code in TaskResource. Let me refactor this part of the send..Response() code to make them reusable. Back to your question. I don't have an answer where this 1024 is from. But I think this size is enough to contain any meaningful error messages and there is not much need to make it configurable.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 3]
```diff
@@ -231,6 +253,25 @@ void PrestoServer::run() {
             .sendWithEOM();
       });
 
+  // Operational endpoints:
+  httpServer_->registerGet(
+      "/v1/operation/connector",
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:258]: Yeah we can do that

File: presto-native-execution/presto_cpp/main/PrestoServer.h [hunk 4]
```diff
@@ -110,6 +112,14 @@ class PrestoServer {
 
   void populateMemAndCPUInfo();
 
+  void runOperation(
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.h:115]:
```
// Invoked to run operation on this server per http request.
```

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 5]
```diff
@@ -552,6 +554,53 @@ void PrestoServer::populateMemAndCPUInfo() {
...
+  try {
+    ServerOperation op = buildServerOpFromHttpRequest(message);
+    switch (op.target) {
+      case ServerOperation::Target::kConnector:
+        http::sendOkResponse(downstream, connectorOperation(op, message));
+        break;
+    }
+  } catch (const VeloxUserError& ex) {
+    http::sendErrorResponse(downstream, ex.what());
+  } catch (const VeloxException& ex) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:569]: Why two catches?

File: presto-native-execution/presto_cpp/main/ServerOperation.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+  Target target;
+  Action action;
+};
+
+/// Builds a server operation from an HTTP request. Throws upon build failure.
+ServerOperation buildServerOpFromHttpRequest(
+    const proxygen::HTTPMessage* httpMsg);
+
+} // namespace facebook::presto
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/ServerOperation.h:56]: Add empty line at the end?

File: presto-native-execution/presto_cpp/main/ServerOperation.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    VELOX_FAIL();
+  }
+  return it->second;
+}
+
+ServerOperation buildServerOpFromHttpRequest(
+    const proxygen::HTTPMessage* message) {
+  const auto& path = message->getPath();
+  auto targetPos = std::string("/v1/operation/").length();
+  auto actionPos = path.find('/', targetPos) + 1;
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/ServerOperation.cpp:55]: Let's also check on targetPos for safe?

File: presto-native-execution/presto_cpp/main/ServerOperation.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  if (it == kReverseActionLookup.end()) {
+    VELOX_FAIL();
+  }
+  return it->second;
+}
+
+ServerOperation buildServerOpFromHttpRequest(
+    const proxygen::HTTPMessage* message) {
+  const auto& path = message->getPath();
+  auto targetPos = std::string("/v1/operation/").length();
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/ServerOperation.cpp:54]:
```
nit: const auto targetPos =

dittos
```

File: presto-native-execution/presto_cpp/main/ServerOperation.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    const std::string& str) {
+  auto it = kTargetLookup.find(str);
+  if (it == kTargetLookup.end()) {
+    VELOX_USER_FAIL("Unsupported server operation target '{}'", str);
+  }
+  return it->second;
+}
+
+std::string ServerOperation::targetString(ServerOperation::Target target) {
+  auto it = kReverseTargetLookup.find(target);
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/ServerOperation.cpp:28]: Let's implement tagetString use a switch case as Target is well defined enum and we can easily handle error using default case? Thanks!

File: presto-native-execution/presto_cpp/main/ServerOperation.cpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+ServerOperation::Action ServerOperation::actionFromString(const std::string& str) {
+  auto it = kActionLookup.find(str);
+  if (it == kActionLookup.end()) {
+    VELOX_USER_FAIL("Unsupported server operation action '{}'", str);
+  }
+  return it->second;
+}
+
+std::string ServerOperation::actionString(Action action) {
+  auto it = kReverseActionLookup.find(action);
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/ServerOperation.cpp:44]: ditto

File: presto-native-execution/presto_cpp/main/ServerOperation.h [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+/// Defines a server operation.
+struct ServerOperation {
+  /// The target this operation is operating upon
+  enum class Target {
+    kConnector,
+  };
+  inline static const std::unordered_map<std::string, Target> kTargetLookup = {
+      {"connector", Target::kConnector}};
+  inline static const std::unordered_map<Target, std::string>
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/ServerOperation.h:29]: kReverseTargetLookup is not required.

File: presto-native-execution/presto_cpp/main/ServerOperation.h [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+  static Target targetFromString(const std::string& str);
+  static std::string targetString(Target target);
+
+  /// The action this operation is trying to take
+  enum class Action { kClearCache, kGetCacheStats };
+  inline static const std::unordered_map<std::string, Action> kActionLookup = {
+      {"clearCache", Action::kClearCache},
+      {"getCacheStats", Action::kGetCacheStats}};
+  inline static const std::unordered_map<Action, std::string>
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/ServerOperation.h:40]: ditto

File: presto-native-execution/presto_cpp/main/ServerOperation.h [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+      {"connector", Target::kConnector}};
+  inline static const std::unordered_map<Target, std::string>
+      kReverseTargetLookup = {{Target::kConnector, "connector"}};
+
+  static Target targetFromString(const std::string& str);
+  static std::string targetString(Target target);
+
+  /// The action this operation is trying to take
+  enum class Action { kClearCache, kGetCacheStats };
+  inline static const std::unordered_map<std::string, Action> kActionLookup = {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/ServerOperation.h:37]: We can move kActionLookup and kTargetLookup to the anonymous namespace in cc
```
s/kActionLookup/kActionLookupMap/
s/kTargetLookup/kTargetLookupMap/
```

File: presto-native-execution/presto_cpp/main/http/HttpConstants.h [hunk 14]
```diff
@@ -14,11 +14,11 @@
 #pragma once
 namespace facebook::presto::http {
 
-const int kHttpOk = 200;
-const int kHttpAccepted = 202;
-const int kHttpNoContent = 204;
-const int kHttpNotFound = 404;
-const int kHttpInternalServerError = 500;
+const uint16_t kHttpOk = 200;
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/http/HttpConstants.h:17]: Nice!

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 15]
```diff
@@ -552,6 +554,53 @@ void PrestoServer::populateMemAndCPUInfo() {
...
+    http::sendErrorResponse(downstream, ex.what());
+  }
+}
+
+std::string PrestoServer::connectorOperation(
+    const ServerOperation& op,
+    proxygen::HTTPMessage* message) {
+  auto name = message->getQueryParam("name");
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:577]: const auto name & id.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 16]
```diff
@@ -552,6 +554,53 @@ void PrestoServer::populateMemAndCPUInfo() {
...
+  auto id = message->getQueryParam("id");
+  if (name == "hive") {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:579]:
```
if (name != "hive") {
   VELOX_USER_FAIL("connector '{}' operation is not supported", name);
}
const auto id = xxx;
the rest of processing...
```

File: presto-native-execution/presto_cpp/main/http/HttpServer.cpp [hunk 17]
```diff
@@ -18,6 +18,55 @@
 
 namespace facebook::presto::http {
 
+void sendOkResponse(proxygen::ResponseHandler* downstream) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/http/HttpServer.cpp:21]: Thanks for the refactoring!

File: presto-native-execution/presto_cpp/main/tests/ServerOperationTest.cpp [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+       lookupIt++) {
+    EXPECT_EQ(
+        ServerOperation::actionFromString(lookupIt->first), lookupIt->second);
+    EXPECT_EQ(ServerOperation::actionString(lookupIt->second), lookupIt->first);
+  }
+  EXPECT_THROW(
+      ServerOperation::actionFromString("UNKNOWN_ACTION"),
+      velox::VeloxUserError);
+}
+
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/tests/ServerOperationTest.cpp:91]: I think we might consider to add PrestoServerTest.cpp to allow us to setup a presto cpp server and run test on top of that as a coordinator in followup.

**Review by xiaoxmeng — APPROVED** [review]
@tanjialiang LGTM. Thanks for update and left some minors.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 5 above)
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:569]: There is some non-user internal error that might be thrown. Just want to differentiate when exceptions got propagated.

File: presto-native-execution/presto_cpp/main/ServerOperation.cpp (hunk 9 above)
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/ServerOperation.cpp:28]: I'm using a lookup/reverselookup mainly to guard the case that developers try to add something and forget to update something. I have a test guarding this. And btw just curious what do you mean by easily handling errors?

File: presto-native-execution/presto_cpp/main/ServerOperation.h (hunk 11 above)
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/ServerOperation.h:29]: Answered in the above comments

File: presto-native-execution/presto_cpp/main/tests/ServerOperationTest.cpp (hunk 18 above)
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/tests/ServerOperationTest.cpp:91]: We sort of have it in Java integration tests

**Review by xiaoxmeng — COMMENTED** [review]
@tanjialiang  thanks for the update!

**Pull request merged**